Feature | Rerun Last RPC

isopropylcyanide opened the pull request:
Implementation for #416 

- Handled server streaming and unary RPC rerun. Client streaming and bidirectional not supported.
- **Slight change from spec:** `call -r` takes the RPC name and hence has the ability to rerun the last request for each different RPC. Earlier, the scope was limited to just the last call. Now it has been extended to the last call per RPC giving us more flexibility.
- Fix minor typos.

**Tested cases**
| Case | Expected |
| --- | --- |
| Old RPC  message was correct and server is running | Should rerun the last command successfully |
| Old RPC message was correct and server is not running | Should fail at `dial()` and run once the server restarts |
| Old RPC message was incorrect / impartial and server is running  | No previous call is recorded if request fails validation |
| Old RPC message was incorrect / impartial and server is not running | No previous call is recorded if request fails validation |

---

**Testing Scenarios (Unary)**
```sh
 ______
 |  ____|
 | |__    __   __   __ _   _ __    ___
 |  __|   \ \ / /  / _. | | '_ \  / __|
 | |____   \ V /  | (_| | | | | | \__ \
 |______|   \_/    \__,_| |_| |_| |___/

 more expressive universal gRPC client


127.0.0.1:5435> package protos

protos@127.0.0.1:5435> service HelloWorld

protos.HelloWorld@127.0.0.1:5435> call SayHello 
message (TYPE_STRING) => Hi there
{
  "message": "Hi there"
}

protos.HelloWorld@127.0.0.1:5435> call --repeat
command call: argument required

protos.HelloWorld@127.0.0.1:5435> call --repeat SayHello
{
  "message": "Hi there"
}

protos.HelloWorld@127.0.0.1:5435> call --repeat SayHello --enrich
content-type: application/grpc

{
  "message": "Hi there"
}

code: OK
number: 0
message: ""

protos.HelloWorld@127.0.0.1:5435> call SayHello
message (TYPE_STRING) => newer
{
  "message": "newer"
}

protos.HelloWorld@127.0.0.1:5435> call SayHello -r
{
  "message": "newer"
}
```

---

**Testing Scenarios (Server Streaming where `num` controls the number of server responses)**
```sh
mirror.MirrorService@127.0.0.1:5435> call MirrorStream
num (TYPE_INT32) => 3
✔ name
name (TYPE_STRING) => Aman
<repeated> f32 (TYPE_FIXED32) => 2
<repeated> f32 (TYPE_FIXED32) => 
what_time::seconds (TYPE_INT64) => 100
what_time::nanos (TYPE_INT32) => 22
{
  "f32": [
    2
  ],
  "name": "Aman",
  "waitTime": 2,
  "whatTime": "1970-01-01T00:01:40.000000022Z"
}
{
  "f32": [
    2
  ],
  "name": "Aman",
  "waitTime": 2,
  "whatTime": "1970-01-01T00:01:40.000000022Z"
}
{
  "f32": [
    2
  ],
  "name": "Aman",
  "waitTime": 2,
  "whatTime": "1970-01-01T00:01:40.000000022Z"
}

mirror.MirrorService@127.0.0.1:5435> call MirrorStream -r --enrich
content-type: application/grpc
{
  "f32": [
    2
  ],
  "name": "Aman",
  "waitTime": 2,
  "whatTime": "1970-01-01T00:01:40.000000022Z"
}
{
  "f32": [
    2
  ],
  "name": "Aman",
  "waitTime": 2,
  "whatTime": "1970-01-01T00:01:40.000000022Z"
}
{
  "f32": [
    2
  ],
  "name": "Aman",
  "waitTime": 2,
  "whatTime": "1970-01-01T00:01:40.000000022Z"
}
rpc-service: mirror

code: OK
number: 0
message: ""

mirror.MirrorService@127.0.0.1:5435> 
```

isopropylcyanide commented on e2e/e2egen/e2egen.go:
```diff
@@ -135,7 +135,7 @@ func generateFile(w io.Writer, src, testCaseName string, input, args []string) e
 		l4 := file.LineStart(fset.Position(nextLine).Line + 4)
 		l5 := file.LineStart(fset.Position(nextLine).Line + 5)
 
-		elts := append(compLit.Elts, &ast.KeyValueExpr{
```
Raised by gocritic. Hence, fixing it.
```
 appendAssign: append result not assigned to the same slice (gocritic)
```

ktr0731 commented:
I'm sorry for the late reply. I'll review it tomorrow, so could you wait until tomorrow? 🙇

ktr0731 commented on usecase/call_rpc.go:
```diff
@@ -416,6 +429,44 @@ func (m *dependencyManager) CallRPC(ctx context.Context, w io.Writer, rpcName st
+		}
+		return nil
+	}
+	return errors.New("cannot rerun previous RPC as client streaming RPCs are not supported")
```
There is no previous command, Evans shows this message. I think it's not expected behavior.
Should we show `No previous request body exists for RPC~` instead of it?

```
$ evans -s -r repl

api.Example@127.0.0.1:50051> call --repeat Unary
command call: cannot rerun previous RPC as client streaming RPCs are not supported 
```

isopropylcyanide replied:
That's absolutely right. I missed this branch.

ktr0731 commented on usecase/call_rpc.go:
```diff
@@ -416,6 +429,44 @@ func (m *dependencyManager) CallRPC(ctx context.Context, w io.Writer, rpcName st
+	if m.state.rpcCallState[id].repeatable {
+		previousReqBytes := m.state.rpcCallState[id].reqPayload
+		if previousReqBytes == nil {
+			return errors.New(fmt.Sprintf("No previous request body exists for RPC: %s. Please issue a normal request.", rpcName))
```
According to [Go CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments#error-strings), we should be:

```suggestion
			return errors.Wrapf("no previous request body exists for RPC: %s, please issue a normal request", rpcName)
```

isopropylcyanide replied:
:D New Gopher here. Thank you for guiding me on the comment standard. Helpful. 
Although I believe there exists no `Wrapf()` that takes in no error. You probably meant `Errorf`)

ktr0731 replied:
> Although I believe there exists no Wrapf() that takes in no error. You probably meant Errorf)

Oops, it's right!

ktr0731 commented on usecase/call_rpc.go:
```diff
@@ -416,6 +429,44 @@ func (m *dependencyManager) CallRPC(ctx context.Context, w io.Writer, rpcName st
 
+// Gets a request with the body containing the payload of its previous RPC
+// Only RPCs that are repeatable by definition will have their previous requests returned.
+func (m *dependencyManager) getPreviousRPCRequest(rpcName string, req interface{}) error {
```
```suggestion
func (m *dependencyManager) getPreviousRPCRequest(rpcName string, req *dynamic.Message) error {
```

ktr0731 commented on usecase/usecase.go:
```diff
@@ -22,14 +22,21 @@ type dependencyManager struct {
+
+type callState struct {
+	reqPayload []byte
+	repeatable bool
```
We can check whether an RPC is ClientStreaming or BidiStreaming by using `*grpc.RPC` (it's available in 
 `CallRPC` method in `usecase/call_rpc.go`. So, I think we don't need to store ClientStreaming/BidiStreaming states.

isopropylcyanide replied:
Good point. It's redundant to store the "repeatable" state. However, I'm still keen on keeping the `callState` struct (even though it has a single field right now) to make it easier to add them later.

ktr0731 replied:
> to make it easier to add them later.

I see. It certainly makes sense, so let's keep it that way!

ktr0731 replied:
I agree with your opinion. It certainly makes sense!

ktr0731 commented on usecase/call_rpc.go:
```diff
@@ -46,21 +50,30 @@ func (e *gRPCError) Code() ErrorCode {
-func (m *dependencyManager) CallRPC(ctx context.Context, w io.Writer, rpcName string, filler fill.Filler) error {
+func (m *dependencyManager) CallRPC(ctx context.Context, w io.Writer, rpcName string, rerunPrevious bool, filler fill.Filler) error {
 	fqsn := proto.FullyQualifiedServiceName(m.state.selectedPackage, m.state.selectedService)
 	rpc, err := m.spec.RPC(fqsn, rpcName)
```
We should return an error if `rerunPrevious` is specified for a ClientStreaming/BidiStreaming RPC. For example,

``` go
if rerunPrevious && rpc.IsClientStreaming {
	return errors.New("cannot rerun previous RPC as client/bidi streaming RPCs are not supported")
}
```
```

isopropylcyanide replied:
This happens in the `getPreviousRPCRequest()`. Didn't want to pollute the `callRPC()` with the responsibility of checking if `isClientStreaming()`. However, as an optimization, we'll save few compute cycles by failing fast. Do you see adding it up top as an advantage?

ktr0731 replied:
Evans has hung up in the previous implementation when calling a BidiStreaming with `--repeat` option.

```
$ evans -s -r repl --package grpc.reflection.v1alpha --service ServerReflection
grpc.reflection.v1alpha.ServerReflection@127.0.0.1:50051> call --repeat ServerReflectionInfo 
# Hung up!
```

The root cause is the `errgroup` which is used for asynchronous processing in the request/response loop (L148-247). We can still keep code without fast-failing, however, it's difficult to keep simplicity.
I think fast-failing is the simplest and most maintainable. What do you think?

isopropylcyanide replied:
Agreed. Failing fast seems reasonable here.

ktr0731 reviewed (commented):
Sorry for the late reply.

The implementation is mostly what I expected. Great work!
I commented on some points. Could you take a look at it when you have a moment?

isopropylcyanide commented on go.mod:
```diff
@@ -39,6 +39,7 @@ require (
 	github.com/spf13/cobra v1.1.3
 	github.com/spf13/pflag v1.0.5
 	github.com/spf13/viper v1.7.1
+	github.com/stretchr/testify v1.7.0
```
This makes it very easy to assert errors. I noticed we were not using this package. Let me know if this is an overhead.

ktr0731 replied:
While Testify is a very powerful tool, it is not a required library for this tool. For this reason, we have adopted a policy of not using Testify in this project.
I am sorry, but could you please replace it with the standard error package? 🙇

isopropylcyanide replied:
**Updated the test to not use testify**. Also, thank you for raising this. I refactored the tests to use the table driven approach which makes them a lot easier to read, debug and reduces duplication. Thanks

isopropylcyanide commented:
Thank you for taking out the time to review this, on a weekday :) I've addressed your comments. Let me know if there's anything else that I may have missed.

ktr0731 reviewed (commented):
Thank you for your quick action.
Just more one step! I have left some replies, so could you check it out when you have time?

isopropylcyanide commented:
Thanks. Will work on the comments this week.

ktr0731 reviewed (approved):
LGTM 👍 👍 
Thank you for your contribution. Evans now gets the powerful feature!

hit9 commented:
Nice feature! Waiting for a release  :)

ktr0731 commented:
@hit9 Hi. We released this feature in a new release, v0.10.0!
